packages/phalanx/src/Phalanx_DataLayout_MDALayout.hpp

```diff
@@ -49,111 +49,60 @@
 #include <string>
 #include <iostream>
 #include <vector>
+#include <type_traits>
 
 namespace PHX {
 
-  // ******************************************************************
-  template<std::size_t, std::size_t> struct
-  Check_num_ctor_arguments_equal_to_num_template_arguments;
+  // ***********************
+  // Pack counting template
+  // ***********************
+  template<typename...ParamPack> struct PackSize;
 
-  // ******************************************************************
-  template<typename Tag0, typename Tag1, typename Tag2, typename Tag,
-	   typename Tag4, typename Tag5, typename Tag6, typename Tag7>
-  struct DLTagList;
+  template<typename...ParamPack>
+  struct PackSize : std::integral_constant<int,PackSize<void,ParamPack...>::value> {};
+
+  template<>
+  struct PackSize<void> : std::integral_constant<int,0> {};
+
+  template<typename NextMember, typename...ParamPack>
+  struct PackSize<void,NextMember,ParamPack...> : std::integral_constant<int,1+PackSize<void,ParamPack...>::value>{};
 
   // ******************************************************************
   /*! \brief A concrete implementation of the DataLayout class for compile time checked multidimensional arrays.
   */
-  template<typename Tag0 = void, typename Tag1 = void, 
-	   typename Tag2 = void, typename Tag3 = void,
-	   typename Tag4 = void, typename Tag5 = void,
-	   typename Tag6 = void, typename Tag7 = void>
-  class MDALayout;
-
-  // ******************************************************************
-  template<typename Tag0, typename Tag1, typename Tag2, typename Tag3,
-	   typename Tag4, typename Tag5, typename Tag6, typename Tag7>
+  template<typename... Tags>
   class MDALayout : public DataLayout {
 
   public:
+    static constexpr int Rank = PHX::PackSize<Tags...>::value;
+    static_assert(Rank > 0);
```

Comment:
I guess another caveat would be calling the `setExtentsVariadic` in the ctor, which currently expects at least one extent. That would also have to be guarded by constexpr if, and possibly the method specialized for no extent. 

Ok, perhaps rank-0 is not that important.

Reply:
I've never seen a requirement for rank 0 and truly hope that is never needed :) DataLayout is intended to just carry a string identifier and the sizes needed to build the multidimensional array. The templated version is one derived version. They could always use the runtime implementation to avoid this. I would like to keep the class as clean and simple as possible. Supporting rank 0 will certainly add complexity as you have pointed out. If we ever do need it, it should be easy to fix.